tests/unit/s2n_tls13_cert_verify_test.c

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+#include "tls/s2n_tls13_certificate_verify.h"
+#include "tls/s2n_cipher_suites.h"
+#include "tls/s2n_tls.h"
+#include "tls/s2n_tls13.h"
+#include "error/s2n_errno.h"
+#include "crypto/s2n_ecdsa.h"
+
+uint8_t hello[] = "Hello, World!\n";
+uint8_t goodbye[] = "Goodbye, World!\n";
+
+struct s2n_tls13_cert_verify_test {
+    const char * const cert_file;
+    const char * const key_file;
+    const struct s2n_signature_scheme *sig_scheme;
+    const s2n_mode sender_mode;
+    const s2n_mode verifier_mode;
+};
+
+const struct s2n_tls13_cert_verify_test test_cases[] = {
+        { .cert_file = S2N_ECDSA_P384_PKCS1_CERT_CHAIN, .key_file = S2N_ECDSA_P384_PKCS1_KEY,
+          .sig_scheme = &s2n_ecdsa_secp256r1_sha256, .sender_mode = S2N_CLIENT,
+          .verifier_mode = S2N_SERVER },
+        { .cert_file = S2N_ECDSA_P384_PKCS1_CERT_CHAIN, .key_file = S2N_ECDSA_P384_PKCS1_KEY,
+          .sig_scheme = &s2n_ecdsa_secp256r1_sha256, .sender_mode = S2N_SERVER,
+          .verifier_mode = S2N_CLIENT },
```

Comment:
add another S2N_RSA_2048_PKCS1_CERT_CHAIN/S2N_RSA_2048_PKCS1_KEY test case here too?

Reply:
Isn't that covered? We do PSS if PSS is supported, and TLS1.3 only allows PSS for RSA certs.

Reply:
But since we do want to test client AND server as the verifier for every test case, you might not need these new fields. Might be worth just automatically testing both directions for every test case.

Reply:
without those fields, i have to duplicate ALL the tests below going in each direction. ~300 lines. And then if there are changes you have to remember to do it in soo many places. And write any new test twice. With the fields, the duplication is contained to 6 lines and its very easy to see what's going. I can add a comment if you think it would help, but I'd like to keep the fields

Reply:
Could you add a "verifier" field to run_tests instead, and make main call run_tests twice?

Reply:
If we are testing against ecdsa and pss certs, I don't see why not to test rsa certs, the effort is also minimal